src/services/rln.ts

```diff
@@ -0,0 +1,147 @@
+  Status = "status",
+  Keystore = "keystore-changed",
+}
+
+export enum StatusEventPayload {
+  WASM_LOADING = "WASM Blob download in progress...",
+  WASM_FAILED = "Failed to download WASM, check console",
+  CONTRACT_LOADING = "Connecting to RLN contract",
+  CONTRACT_FAILED = "Failed to connect to RLN contract",
+  RLN_INITIALIZED = "RLN dependencies initialized",
+  KEYSTORE_LOCAL = "Keystore initialized from localStore",
+  KEYSTORE_NEW = "New Keystore was initialized",
+  CREDENTIALS_REGISTERING = "Registering credentials...",
+  CREDENTIALS_REGISTERED = "Registered credentials",
+  CREDENTIALS_FAILURE = "Failed to register credentials, check console",
+}
+
+type EventListener = (event: CustomEvent) => void;
+
+type IRLN = {
+  saveKeystore: () => void;
+  addEventListener: (name: RLNEventsNames, fn: EventListener) => void;
+  removeEventListener: (name: RLNEventsNames, fn: EventListener) => void;
+};
+
+export class RLN implements IRLN {
+  private readonly emitter = new EventTarget();
+  public readonly ethProvider: ethers.providers.Web3Provider;
+
+  public rlnInstance: undefined | RLNInstance;
+  public rlnContract: undefined | RLNContract;
+  public keystore: Keystore;
+
+  private initialized = false;
+  private initializing = false;
+
+  public constructor() {
+    const ethereum =
+      window.ethereum as unknown as ethers.providers.ExternalProvider;
+    if (!isBrowserProviderValid(ethereum)) {
+      throw Error(
+        "Invalid Ethereum provider present on the page. Check if MetaMask is connected."
+      );
+    }
+    this.ethProvider = new ethers.providers.Web3Provider(ethereum, "any");
+    this.keystore = this.initKeystore();
+  }
+
+  public async init(): Promise<void> {
+    if (this.initialized || this.initializing) {
+      return;
+    }
+
+    this.initializing = true;
+    const rlnInstance = await this.initRLNWasm();
+    await this.initRLNContract(rlnInstance);
+
+    this.emitStatusEvent(StatusEventPayload.RLN_INITIALIZED);
+
+    // emit keystore keys once app is ready
+    this.emitKeystoreKeys();
+
+    this.initialized = true;
+    this.initializing = false;
+  }
+
+  private async initRLNWasm(): Promise<RLNInstance> {
+    this.emitStatusEvent(StatusEventPayload.WASM_LOADING);
+    try {
+      this.rlnInstance = await create();
+      return this.rlnInstance;
+    } catch (error) {
+      console.error(
+        "Failed at fetching WASM and creating RLN instance: ",
+        error
+      );
+      this.emitStatusEvent(StatusEventPayload.WASM_FAILED);
+      throw error;
+    }
+  }
+
+  private async initRLNContract(rlnInstance: RLNInstance): Promise<void> {
+    this.emitStatusEvent(StatusEventPayload.CONTRACT_LOADING);
+    try {
+      this.rlnContract = await RLNContract.init(rlnInstance, {
+        registryAddress: SEPOLIA_CONTRACT.address,
+        provider: this.ethProvider.getSigner(),
+      });
+    } catch (error) {
+      console.error("Failed to connect to RLN contract: ", error);
```

Comment:
I'm getting this error, can we fix that? Running it as explained in the docs

Reply:
Could solve it with connecting manually to the website, but the UX fo this would be bad. Can we have a "connect" button to connect the wallet, or something like that? Also show the connected wallet address (if any)